### DISCUSSION TITLE
armbian

### DISCUSSION THREAD

**Discussion by xlazom00** [jglathe/linux_ms_dev_kit#8] [General]
I am working armbian MR
https://github.com/xlazom00/build/tree/wdk2023
How did you manage to improve wifi firmware? Do I need to add some fw blob?

wdk firmware are there already
https://github.com/armbian/firmware/pull/71

**Comment by xlazom00** [comment]
And I am also thinking if I should just use https://github.com/steev/linux/ repo and your patches  as it will be just less maintenance.

**Comment by jglathe** [comment]
Wow didn't see the discussion until now. Sorry for the delay. Didn't look at armbian for a while, got a bit distracted by other things :grin: Anyway. I'm actually using 23.10 on all devices, with restored thermal management and 6.7-rc5. Trying to get debuild to do what I want, but I'm not optimistic about that. 
WLAN firmware: The firmware itself is unchanged, only the board file is kinda hacked. I tried all available combinations of the existing board file to come up with the best-matching calibration data from the Thinkpad X13s (which would have been the obvious choice anyway). Trying to extract the wdk calibration data from the Windows driver package proved ineffective. 
The resulting file and install hints can be found [here](https://github.com/jglathe/linux_ms_dev_kit/issues/7#issuecomment-1831199212) . The discovery story can be found [here](https://github.com/jglathe/linux_ms_dev_kit/discussions/1#discussioncomment-7475082) . And the next image (probably 6.7) will have it all in one package I guess. 
The wdk builds base on steev's repo for quite a while now I guess. Will upstream the findings re tsens to get it fixed, hope they accept that I don't understand it but can make it work :man_shrugging: 
As for exclusive linux on the local SSD, beware of the seeming hardware dependency regarding thermal management. Still investigating this.

**Comment by jglathe** [comment]
Now that the armbian PR is merged, I have added a few [tweaks](https://github.com/armbian/build/pull/6091). This should bring the armbian builds on par with what I'm using, but sorta more elegant :grin: Anyway. Thank you for doing this.

**Comment by jglathe** [comment]
The tweaks PRs are merged now, too.

**Reply by likeadragonmaid** [reply]
How to install Armbian to the SSD. After I install it to SSD, grub does not gets installed to SSD. Manually installing it gives some UEFI incompatibility issues.

**Reply by jglathe** [reply]
efivars are not (yet) supported, therefore the warnings you see. Doesn't prevent grub from working, though. This is my SSD in GParted:

![Screenshot from 2024-01-30 06-45-40](https://github.com/jglathe/linux_ms_dev_kit/assets/33185912/e12e02df-c79f-42e2-8cbd-b8c3c68f3891)

Note the partition labels. I use them to mount the right ones, even if there's a USB stick inserted.

![Screenshot from 2024-01-30 06-50-45](https://github.com/jglathe/linux_ms_dev_kit/assets/33185912/fec6dc73-8cc9-4aa5-a1c5-9fb182ad12aa)

The labels need to differ from those on the USB stick, highly recommended. Also, install on ssd requires that you install grub on it (with chroot) 

```
#mount the ssd file on /mnt
#device names for my ssd partitioning, you might need to adapt it
sudo mount  /dev/nvme0n1p5 /mnt/ 
#mount the EFI partition on /mnt/boot/efi
sudo mount /dev/nvme0n1p1 /mnt/boot/efi/
#prepare for chroot
for i in /dev /dev/pts /proc /sys /run; do sudo mount -B $i /mnt/$i; done
#chroot
sudo chroot /mnt/

#install grub
apt-get install grub2-common grub-efi
grub-install

#check that /etc/default/grub has the right content

------------------------------------------------------------------
# If you change this file, run 'update-grub' afterwards to update
# /boot/grub/grub.cfg.
# For full documentation of the options in this file, see:
#   info -f grub -n 'Simple configuration'

GRUB_DEFAULT=saved
GRUB_SAVEDEFAULT=true
#GRUB_TIMEOUT_STYLE=hidden
GRUB_TIMEOUT=5
GRUB_DISTRIBUTOR=`lsb_release -i -s 2> /dev/null || echo Debian`
GRUB_CMDLINE_LINUX_DEFAULT="pd_ignore_unused clk_ignore_unused arm64.nopauth efi=noruntime"
GRUB_CMDLINE_LINUX=""

# Uncomment to enable BadRAM filtering, modify to suit your needs
# This works with Linux (no patch required) and with any kernel that obtains
# the memory map information from GRUB (GNU Mach, kernel of FreeBSD ...)
#GRUB_BADRAM="0x01234567,0xfefefefe,0x89abcdef,0xefefefef"

# Uncomment to disable graphical terminal (grub-pc only)
#GRUB_TERMINAL=console

# The resolution used on graphical terminal
# note that you can use only modes which your graphic card supports via VBE
# you can see them in real GRUB with the command `vbeinfo'
#GRUB_GFXMODE=640x480

# Uncomment if you don't want GRUB to pass "root=UUID=xxx" parameter to Linux
#GRUB_DISABLE_LINUX_UUID=true

# Uncomment to disable generation of recovery mode menu entries
#GRUB_DISABLE_RECOVERY="true"

# Uncomment to get a beep at grub start
#GRUB_INIT_TUNE="480 440 1"

#enable os-prober for better use with Windows on the internal nvme
GRUB_DISABLE_OS_PROBER=false
------------------------------------------------------------------

#run update-grub
update-grub
# run update-initramfs -u
update-initramfs -u -k all
#Done!
exit
for i in /mnt/dev /mnt/dev/pts /mnt/proc /mnt/sys /mnt/run; do sudo umount -l $i; done
sudo umount /dev/nvme0n1p1
sudo umount /dev/nvme0n1p5

```
This should do it. armbian is odd in some ways, I normally don't use it. But it should do.

**Reply by jglathe** [reply]
efivar is also supported now, with [this](https://github.com/jglathe/linux_ms_dev_kit/commit/5bde928e8154c86b9ff78f5385131a321c81a3d1) commit.

**Comment by xlazom00** [comment]
@jglathe Can you plz update armbian configurations to latest kernel ?
It will be  nice to have latest kernel
Current 6.7.0-rc6 is kind of old :)

**Reply by xlazom00** [reply]
@jglathe Should I use 
https://github.com/jglathe/linux_ms_dev_kit/commits/jg/blackrock-v6.8.y/
or
https://github.com/jglathe/linux_ms_dev_kit/tree/jg/mainline
?

**Comment by jglathe** [comment]
Mainline is on 6.8.7 as of now, it is rebased on Ubuntu Mainline, therefore the different branch name. Newest is [6.9-rc5](https://github.com/jglathe/linux_ms_dev_kit/tree/jg/v6.9-rc5).

**Comment by jglathe** [comment]
Need to update the fw_fetcher and syshacks repos with newest developments. And yes, an updated armbian build script might be due.

**Reply by mirostauder** [reply]
Updated my armbian kernel build script,
https://github.com/jglathe/linux_ms_dev_kit/discussions/10#discussioncomment-8532326
I am running on `jg/v6.9-rc5` now,
```
Linux wdk2023 6.9.0-rc5-g67d03c759b1e #1 SMP PREEMPT Fri Apr 26 14:51:00 UTC 2024 aarch64 GNU/Linux
```
so far so good!

**Reply by mirostauder** [reply]
on a second look:
```
root@wdk2023: /root # dmesg | grep -i failed
[    0.000000] psci: [Firmware Bug]: failed to set PC mode: -3
[    3.087857] pmic_glink: module verification failed: signature and/or required key missing - tainting kernel
[    3.237832] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a600000.usb
[    3.239840] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a800000.usb
[    3.650564] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with 88eb000.phy
[    3.661522] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with 8903000.phy
[    3.715474] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a600000.usb
[    3.756506] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a800000.usb
[    9.186303] energy_model: Accessing cpu4 policy failed
[   19.681797] snd-sc8280xp sound: ASoC: failed to instantiate card -110
[   19.682063] snd-sc8280xp sound: probe with driver snd-sc8280xp failed with error -110
```

**Reply by mirostauder** [reply]
after `apt full-upgrade` and reboot:
```
root@wdk2023: /root # dmesg | grep -i fail
[    0.000000] psci: [Firmware Bug]: failed to set PC mode: -3
[    2.940885] pmic_glink: module verification failed: signature and/or required key missing - tainting kernel
[    3.524996] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with 88eb000.phy
[    3.526301] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with 8903000.phy
[    3.570554] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a600000.usb
[    3.611815] qcom_pmic_glink pmic-glink: Failed to create device link (0x180) with a800000.usb
[   10.037149] energy_model: Accessing cpu4 policy failed
[   20.964890] platform 33c0000.pinctrl: deferred probe pending: qcom-sc8280xp-lpass-lpi-pinctrl: Failed to get clk 'core'
[   27.362172] snd-sc8280xp sound: ASoC: failed to instantiate card -110
[   27.362478] snd-sc8280xp sound: probe with driver snd-sc8280xp failed with error -110
```
this is:
```
oot@wdk2023: /root # cat /etc/os-release 
PRETTY_NAME="Armbian 24.5.0-trunk.489 trixie"
NAME="Debian GNU/Linux"
VERSION_CODENAME=trixie
ID=debian
HOME_URL="https://www.armbian.com/"
SUPPORT_URL="https://forum.armbian.com"
BUG_REPORT_URL="https://www.armbian.com/bugs"
ARMBIAN_PRETTY_NAME="Armbian 24.5.0-trunk.489 trixie"
```

**Comment by jglathe** [comment]
Hmm these errors are sorta "normal". Sound is not complete, and pmic glink is clashing a little with UCSI. There were some commits where the pmic glink errors went away, but they're still on it. There are new patches for sound and soundwire support on DP, maybe there's a chance to tackle this now. Did you try the EL2 branch, just out of curiosity? I'm sort of constantly running it on one box.
